Standardized pod security profiles

tallclair opened the pull request:
Document a standard set of pod security policies targeting a wide set of workloads. This is a markdown conversion of [this document](https://docs.google.com/document/d/1d9c4BaDzRw1B5fAcf7gLOMZSVEvrpSutivjfNOwIqT0/edit#heading=h.ihgl7qlgtyzu) with some small content tweaks. This was presented to [SIG-Auth on February 19th](https://docs.google.com/document/d/1woLGRoONE3EBVx-wTb4pvp4CI7tmLZ6lS26VTbosLKM/edit#heading=h.bxqchvtk3vmr), and was well-received.

The goal is to be able to provide a set of standard implementation-agnostic policy recommendations. This will help users better understand security best practices, will help security benchmarks like CIS provide better recommendations, and our hope is that it will move the Kubernetes ecosystem (e.g. helm charts) towards a more standard set of permissions.

/sig auth

tallclair commented on content/en/docs/concepts/policy/pod-security-standards.md:
```diff
@@ -0,0 +1,323 @@
+{{% capture overview %}}
+
+Security settings for pods are typically applied by using [security
+contexts](../../tasks/configure-pod-container/security-context/). Security Contexts allows for the
```
I seem to be missing an extra level of `../` on all the relative links. I'll add it tomorrow.

tallclair replied:
Ah, thanks. I hadn't seen the style guide, and assumed

sftim commented on content/en/docs/concepts/policy/pod-security-standards.md:
```diff
@@ -0,0 +1,323 @@
+definition of privilege and access controls on a per-pod basis.
+
+The enforcement and policy-based definition of cluster requirements of security contexts has
+previously been achieved using [Pod Security Policy](pod-security-policy/). A _Pod Security Policy_
```
Link to `/docs/concepts/policy/pod-security-policy/`

sftim commented on content/en/docs/concepts/policy/pod-security-standards.md:
```diff
@@ -0,0 +1,323 @@
+developers of non-critical applications. The following listed controls should be
+enforced/disallowed:
+
+<table>
```
Can you use a Markdown table here? If you can, also add a [caption](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#table-captions) using a shortcode; if not, please add an equivalent caption manually.

tallclair replied:
Unfortunately not, because of the nested lists (at least I couldn't figure out how to make it work). I copied the caption HTML from the rendered table in the style guide.

sftim commented on content/en/docs/concepts/policy/pod-security-standards.md:
```diff
@@ -0,0 +1,323 @@
+
+Windows in Kubernetes has, by default, some limitations and differentiators from standard
+Linux-based workloads. Specifically, none of the PodSecurityContext fields [work on
+Windows](../../setup/production-environment/windows/intro-windows-in-kubernetes/#v1-podsecuritycontext). As
```
Link to `/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#v1-podsecuritycontext`

tallclair commented:
Thanks for the review. Moved to security (I wasn't sure where to put it, but that sounds good to me)

jimangel commented:
Love this doc and idea!

Preview: https://deploy-preview-20789--kubernetes-io-master-staging.netlify.app/docs/concepts/security/pod-security-standards/

I agree with @sftim about this living in security, but should we move or combine the PSP document located at: https://deploy-preview-20789--kubernetes-io-master-staging.netlify.app/docs/concepts/policy/pod-security-policy/

At the very least potentially link the two docs?

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+
+Security settings for pods are typically applied by using [security
+contexts](/docs/tasks/configure-pod-container/security-context/). Security Contexts allows for the
+definition of privilege and access controls on a per-pod basis.
```
Hello @tallclair .
A few nits:
s/pods/Pods/g
s/pod/Pod/g

Security Contexts allow?

Line 16: Move `A _Pod Security` to the next line?

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+preventing known privilege escalations. This policy is targeted at application operators and
+developers of non-critical applications. The following listed controls should be
+enforced/disallowed:
+
```
Note: There is a `table` shortcode for adding a table caption.
https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#table-captions

kbhawkey replied:
The table is rendering oddly (styling of lists?). I have not investigated the details.
I _think_ you can include HTML lists in a Markdown table.
It would take time to convert the table to Markdown.

tallclair replied:
Hmm, the styling has to do with CSS rules that are applied to `<p>` and `<li>` elements overriding the styling rules for table content, hence the mixed sizes. This rendered a lot better in github markdown.

I see 2 options, LMK if you can thinking of anything else:
1. Scrap the tables, just make the control column into section headers, and put everything into normal paragraphs and lists.
2. Add some custom CSS rules to override the styles for the nested elements to make them more consistent with the table style.
3. Add some custom CSS to override the table styles to make them more consistent with paragraph styles (why is the table font size so small anyway?)

I'm leaning towards option 1 since this page is already a big mess of custom HTML and that would clean it up. It's too bad though, since I think the table format is easier to digest.

kbhawkey replied:
@tallclair ,
The content seems to work well with tables.
The docs site is in the process of adding a new theme (new parser recently added) and with that some styling changes. The default styling of tables needs to be addressed.
In the meantime, could you create the table with table data/row elements (column span if needed)?
I totally agree with you about the table data font size.

tallclair replied:
> In the meantime, could you create the table with table data/row elements (column span if needed)?

Sorry, could you clarify? I'm not sure what you mean here.

kbhawkey replied:
Could you create the table without the lists? 
The display may improve.
Styling (from bootstrap) would help visually, though I am hesitant to add styling (table border, border for rows?) just for this table.
```<div class="table-responsive">
   <table class="table table-bordered"><thead class="thead-light">
```

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+
+**PodSecurityPolicy**
+
+*   [Privileged](https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/policy/privileged-psp.yaml)
```
nit: You can remove the extra spaces at start of list item text (lines 274-276)

tallclair replied:
Done. (It was an artifact of the automatic conversion from a google doc to markdown).

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+niche. This is not to say that the privileged profile should always be used in this case, but that
+policies in this space need to be defined on a case-by-case basis.
+
+We may reconsider this position in the future, should a clear need for other profiles arise.
```
nit: s/We/"Sig-Auth?"/

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+### What's the difference between a security policy and a security context?
+
+Security contexts are sets of conditions relating to how a container is configured that they want to
+be true in order for the container or pod to run. These security conditions are known as [Security
```
nit: possible reword,
Security contexts are sets of conditions relating to how a container is configured.
These contexts [or conditions] have to be true in order for the container or Pod to run.
OR
Security contexts are sets of conditions relating to how a container is configured which must be 
true for the container or Pod to run.

tallclair replied:
I rewrote this paragraph.

kbhawkey replied:
👍

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+Security profiles however are control plane mechanisms to enforce standards relating to these
+security contexts. As of February 2020, the current native solution for enforcing these security
+policies is [Pod Security Policy](/docs/concepts/policy/pod-security-policy/) - a mechanism for centrally enforcing
+security control aspect standards across a cluster - which is currently in Beta. Other alternatives
```
nit: Probably would not mention `Beta`. Is the feature state documented in the policy page?

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+Gatekeeper](https://github.com/open-policy-agent/gatekeeper)
+
+### What profiles should I apply to my windows pods?
+
```
nit: s/windows/Windows/ ?

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+
+There is not currently an API standard that controls whether a pod is considered sandboxed or
+not. However, contributing factors can be used to determine if a pod is flagged - such as the
+defined runtime of gvisor, kata etc). These contributing factors can be used collectively to provide
```
nit: missing beginning paren?

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+
+Windows in Kubernetes has, by default, some limitations and differentiators from standard
+Linux-based workloads. Specifically, none of the PodSecurityContext fields [work on
+Windows](/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#v1-podsecuritycontext). As
```
possible reword:
Specifically, the PodSecurityContext fields have no effect on Windows?

tallclair commented:
Thanks for the feedback @kbhawkey. I reworded a few sections in the FAQ, but left an open question on what to do about the table formatting.

PTAL

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+### Why isn't there a profile between privileged and default?
+
+The 3 profiles defined here have a clear linear progression from most secure (restricted) to least
+secure (privileged), and cover a broad set of workloads. Privileges required above the baseline
```
nit: s/3/three/

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+### What's the difference between a security policy and a security context?
+
+[Security Contexts](/docs/tasks/configure-pod-container/security-context/) configure Pods and
+Containers at runtime. The are defined as part of the Pod and container specifications in the Pod
```
nit: Line 294, The are?

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,325 @@
+manifest, and represent parameters to the container runtime.
+
+Security policies are control plane mechanisms to enforce specific settings in the Security Context,
+as well as other parameters outside the Security Contex.  As of February 2020, the current native
```
nit: extra space before `As`

tallclair commented:
OK, I removed most of the inner HTML on the table cells (mainly `<p>` and `<ul>` elements). I added a bit of styling to separate the sections. I think it looks a lot better, although it's too bad the field paths don't render in monospace.

PTAL.

kbhawkey commented:
@tallclair . Thanks for the table changes.
A nit: I would rather see the variables/fields appear without the dash prefix.
I rewrote the table with just `<tr>` and `<td>` elements. 

<table>
    <caption style="display:none">Baseline policy specification</caption>
    <tbody>
		<tr><td><strong>Control</strong></td><td><strong>Policy</strong></td></tr>
		<tr><td>Host Namespaces</td><td>Sharing the host namespaces must be disallowed.</td></tr>
				<tr><td></td><td><b>Restricted Fields:</b></td></tr>
				<tr><td></td><td>spec.hostNetwork</td></tr>
				<tr><td></td><td>spec.hostPID</td></tr>
				<tr><td></td><td>spec.hostIPC</td></tr>
				<tr><td></td><td><b>Allowed Values:</b> false</td><tbr>
			</td>
		<tr><td>Privileged Containers</td><td>Privileged Pods disable most security mechanisms and must be disallowed.</td>
        <tr><td></td><td><b>Restricted Fields:</b></td></tr>
		<tr><td></td><td>spec.containers[*].securityContext.privileged</td></tr>
		<tr><td></td><td>spec.initContainers[*].securityContext.privileged</td></tr>
		<tr><td></td><td><b>Allowed Values:</b> false, undefined/nil</td></tr>
		<tr><td>Capabilities</td><td>
				Adding additional capabilities beyond the <a href="https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities">default set</a> must be disallowed.</td></tr>
				<tr><td></td><td><b>Restricted Fields:</b></td></tr>
				<tr><td></td><td>spec.containers[*].securityContext.capabilities.add</td></tr>
				<tr><td></td><td>spec.initContainers[*].securityContext.capabilities.add</td></tr>
				<tr><td></td><td><b>Allowed Values:</b> empty (optionally whitelisted defaults)</td></tr>
		<tr><td>HostPath Volumes</td><td>HostPath volumes must be forbidden.</td></tr>
			<tr><td></td><td><b>Restricted Fields:</b></td></tr>
			<tr><td></td><td>spec.volumes[*].hostPath</td></tr>
			<tr><td></td><td><b>Allowed Values:</b> undefined/nil</td></tr>
		<tr><td>Host Ports</td><td>HostPorts should be disallowed, or at minimum restricted to a whitelist.</td>
        </tr>
		<tr><td></td><td><b>Restricted Fields:</b></td></tr>
		<tr><td></td><td>spec.containers[*].ports[*].hostPort</td></tr>
		<tr><td></td><td>spec.initContainers[*].ports[*].hostPort</td></tr>
		<tr><td></td><td><b>Allowed Values:</b> 0, undefined, (whitelisted)</td></tr>
        <tr><td>AppArmor <em>(optional)</em></td><td>On supported hosts, the `runtime/default` AppArmor profile is applied by default. The default policy should prevent overriding or disabling the policy, or restrict overrides to a whitelisted set of profiles.</td>
        </tr>
		<tr><td></td><td><b>Restricted Fields:</b></td></tr>
		<tr><td></td><td>metadata.annotations['container.apparmor.security.beta.kubernetes.io/*']</td></tr>
		<tr><td></td><td><b>Allowed Values:</b> runtime/default, undefined</td></tr>
        <tr><td>SELinux <em>(optional)</em></td><td>Setting custom SELinux options should be disallowed.</td></tr>
		<tr><td></td><td><b>Restricted Fields:</b></td></tr>
		<tr><td></td><td>spec.securityContext.seLinuxOptions</td></tr>
		<tr><td></td><td>spec.containers[*].securityContext.seLinuxOptions</td></tr>
		<tr><td></td><td>spec.initContainers[*].securityContext.seLinuxOptions</td></tr>
		<tr><td></td><td><b>Allowed Values:</b> undefined/nil</td></tr>
	</tbody>
</table>

tallclair commented:
IMO it's harder to read with a separate row for each line. I'm open to removing the faux-bullet `-` though.

kbhawkey commented on content/en/docs/concepts/security/pod-security-standards.md:
```diff
@@ -0,0 +1,300 @@
+Windows in Kubernetes has some limitations and differentiators from standard Linux-based
+workloads. Specifically, the Pod SecurityContext fields [have no effect on
+Windows](/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#v1-podsecuritycontext). As
+such, no standardised Pod Security profiles currently exists.
```
last nit: s/standardised/standardized/

kbhawkey commented:
@tallclair . Changes look good. I noted one nit.
If this PR is ready to go, would you squash your changes into a single commit?

tallclair commented:
Thanks for your feedback @kbhawkey 

Squashed & good to go!

kbhawkey commented:
Thanks @tallclair .
/lgtm

kbhawkey commented:
/approve
